### ISSUE TITLE
T bit on R_ARM relocations with odd addend is handled incorrectly

### ISSUE THREAD

**Issue by john-brawn-arm** [llvm/llvm-project#63390]
Many R_ARM relocations are defined in https://github.com/ARM-software/abi-aa/blob/main/aaelf32/aaelf32.rst as performing the operation `(S + A) | T` where
- S is the address of the symbol.
- A is the addend for the relocation.
- T is 1 if the target symbol S has type STT_FUNC and the symbol addresses a Thumb instruction; it is 0 otherwise.

With the value of symbols having some special rules:
 - If the symbol addresses an Arm instruction, its value is the address of the instruction (in a relocatable object, the offset of the instruction from the start of the section containing it).
 - If the symbol addresses a Thumb instruction, its value is the address of the instruction with bit zero set (in a relocatable object, the section offset with bit zero set).
 - For the purposes of relocation the value used shall be the address of the instruction (st_value & ~1).

This means that if we have a function whose first instruction is a thumb instruction at 0x1000, then if we have a relocation with addend 1 the final relocated value should be `(0x1000 + 1) | 1` i.e. 0x1001. However lld will currently calculate a value of 0x1002.

An example where this happens:
```
 .text
 .thumb_func
fn:
 bx lr

 .data
var1:
 .word fn
var2:
 .word fn+1
```
build with
```
llvm-mc -filetype=obj -triple=armv7a tmp.s -o tmp.o
ld.lld tmp.o -o a.out
llvm-objdump --syms --reloc --full-contents a.out
```
llvm-objdump output has
```
SYMBOL TABLE:
000200d4 l     F .text  00000000 fn
Contents of section .data:
 300d6 d5000200 d6000200                    ........
```
fn starts at address 0x200d4, var1 has value 0x200d5 (correct) and var2 has value 0x200d6 (incorrect, should be 0x200d5).

Looking at lld it's not obvious to me how this should be fixed. It looks like this is happening because the way things are defined in the abi is "remove bit 0 at start, add the addend, then add bit 0 back in at the end" whereas what lld is doing is "take the symbol value as-is and add the addend". It looks like ARM::relocate is too late to do anything, as that's passed the S+A value from TargetInfo::relocateAlloc which got it from InputSectionBase::getRelocTargetVA which got it from Symbol::getVA.

**Comment by smithp35** [comment]
I have been aware of this when the Arm backend first got implemented. At the time the Arm backend was new and getting major changes to LLD to support it required a lot of justification that the change was worthwhile. At the time I couldn't see any compiler that generated code like that, and I couldn't find any instances of assembly code that used it so I didn't think that was a fight I'd be able to win without a real example. IIUC the use case would be cooperating with code that rolls its own interworking.

I did later run into an example from Morello which similarly uses the bottom-bit, but this was again someone rolling their own interworking before tool support and it was simple to change.

Do we have a concrete use case that would justify the change to accomodate it? I didn't think I had one at the time, but you may have found one?

**Comment by smithp35** [comment]
Yes I agree that a comment would be useful, may even be able to that with the R_ARM_THM_ABS_Gn patch. If/when LLD gets a full set of documentation it would also be useful to put the limitation there too. We don't have the equivalent of GNU ld's target specific docs.

**Comment by blandcr** [comment]
I think this issue also causes differing behavior in linker script `ABSOLUTE(someThumbFn)` evaluation between GNU ld and lld. In GNU ld the thumb bit is stripped while in lld it is not. For my own specific use case (satisfying LPC11U37 vectors checksum) the difference does not matter but it certainly surprised me when I noticed it.
